docs/docs/tutorial/creating-a-source-plugin/part-4/index.mdx

```diff
@@ -0,0 +1,865 @@
+    exports.pluginOptionsSchema = ({ Joi }) => {
+      return Joi.object({
+        message: Joi.string()
+          .required()
+          .description(`The message logged to the console.`),
+      })
+    }
+    ```
+
+Here are some tips when working with plugin options:
+
+- **Consider the options a public API contract between your users and your plugin.** Follow [SemVer](https://semver.org/) versioning when adding, modyfing, and removing options. For example, if you want to remove an option you should only do this in a major release of your plugin since it's a breaking change.
+- **Add as little plugin options as possible.** In the beginning, use safe defaults where possible (to the best of your knowledge) and when users wish to have new options available ask yourself really hard if you should add them. Once added, you'll need to maintain the option forever (or until you remove it again).
+- **Use clear names.** Naming things is hard, but take extra care in using clear names for your plugin's options. Best case scenario is that an option is clear enough without any additional description.
+- **Use "enable" options, not negated options.** An example: Instead of using `noWarnings: true`, use `warnings: false`. The double negation that would occur with `noWarnings: false` is really confusing.
+- **Pass options through.** If your source plugin is using a library under the hood (e.g. an SDK from a CMS) that has options on its own, allow people to configure all those options, too. You could collect those under a new key in the `options` object (e.g. `options.sdkOptions`) and then refer people to the SDK's documentation.
+
+### Task: Use `endpoint` option
+
+Instead of hardcoding the URL to the GraphQL API, you'll use an `endpoint` option and thus make the URL configurable.
+
+1. The boilerplate you cloned at the beginning already has an `endpoint` option defined in the site's `gatsby-config`. Go to `site/gatsby-config.ts` to inspect the code:
+
+    ```ts:title=site/gatsby-config.ts
+    import type { GatsbyConfig } from "gatsby"
+    import type { IPluginOptions } from "plugin"
+
+    // highlight-next-line
+    const GRAPHQL_ENDPOINT = `http://localhost:4000/graphql`
+
+    const config: GatsbyConfig = {
+      graphqlTypegen: true,
+      plugins: [
+        {
+          resolve: `plugin`,
+          options: {
+            // highlight-next-line
+            endpoint: GRAPHQL_ENDPOINT,
+          } satisfies IPluginOptions,
+        },
+        `gatsby-plugin-image`,
+        `gatsby-plugin-sharp`,
+        `gatsby-transformer-sharp`,
+      ],
+    }
+
+    export default config
+    ```
+
+    So the plugin already receives the GraphQL endpoint through the `endpoint` option.
+
+1. Update the TypeScript types inside `plugin/src/types.ts` so that both internally and externally the types are correct:
+
+    ```ts:title=plugin/src/types.ts
+    // Rest of types
+
+    interface IPluginOptionsKeys {
+      // highlight-next-line
```

Comment:
Since these instructions are to replace the existing interface by this name, what do you think about highlighting the whole block? Specifically lines 671 - 674.

Reply:
The interface should stay as-is, only the contents of it should change. I'll change the sentence for it a little bit to make that clearer.